### PR TITLE
Implement configurable data migration using SDK search API

### PR DESCRIPTION
**Pull request by ellykits**
Example use case: Fix an issue with extraction for instance this can be used to fix incorreclty extract reference. Configuration accepts data queries and resource type to search then applies the update values on the resources. This function should be blocking and should run immediately after login.

**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes [link to issue]

**Engineer Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 
- [ ] I have checked that this PR does NOT introduce **breaking changes** that require an update to **_Content_** and/or **_Configs_**? _If it does add a sample here or a link to exactly what changes need to be made to the content._

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

**Comment by ellykits** [comment]
@dubdabasoduba I have drafted this PR for your review on how the supposed configurable data migration should work. I updated the value section to use a `RuleConfig` instead for more flexibility. This should allow us to populate the rules facts with more data if needed. I noted this functionality may consume a lot of resources because of the DB reads & writes thus it should only be run once and versioned to prevent unnecessary migration.

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/PreferenceDataStore.kt [hunk 1]
```diff
@@ -51,9 +52,8 @@ class PreferenceDataStore @Inject constructor(@ApplicationContext val context: C
   }
 
   companion object Keys {
-    val appIdKeyName = "appId"
-    val langKeyName = "lang"
-    val APP_ID by lazy { stringPreferencesKey(appIdKeyName) }
-    val LANG by lazy { stringPreferencesKey(langKeyName) }
+    val APP_ID by lazy { stringPreferencesKey("appId") }
+    val LANG by lazy { stringPreferencesKey("lang") }
+    val MIGRATION_VERSION by lazy { intPreferencesKey("migrationVersion") }
```
[the remark this answers is not in the document]
Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/PreferenceDataStore.kt:57]: We can't use `const` keywords on delegated properties. The variables are declared in the `companion object` of the class; thus they are always static and declared only once. The `strings` are only useful in that context and are not used anywhere else in the application. This improves readability.

File: android/quest/src/main/java/org/smartregister/fhircore/quest/data/DataMigration.kt [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  suspend fun migrate(migrationConfigs: List<MigrationConfig>?, previousVersion: Int) {
+    val maxVersion = migrationConfigs?.maxOfOrNull { it.version } ?: previousVersion
+    migrationConfigs?.forEach { migrationConfig ->
+      try {
+        val search: Search =
+          Search(type = migrationConfig.resourceType).apply {
+            migrationConfig.dataQueries?.forEach { dataQuery ->
+              filterBy(dataQuery = dataQuery, configComputedRuleValues = emptyMap())
+            }
+          }
```
[the remark this answers is not in the document]
Comment (ellykits) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/DataMigration.kt:134]: I will add the documentation. Queries used in this contexts are not related to base/related resources but rather any kind of resource stored in the database.

File: android/quest/src/main/java/org/smartregister/fhircore/quest/data/DataMigration.kt [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+   *
+   * Valid expressions (format in key, value -> description) include:
+   * ```
+   * "$.name[0].use", "casual" -> Update the first name usage from "official" to "casual"
+   * "Patient.gender", "male" -> Update the gender from "female" to "male"
+   * "$.birthDate", "1996-12-32" -> Update the birth date from "1974-12-25" to "1996-12-32"
+   * ```
+   */
+  suspend fun migrate(migrationConfigs: List<MigrationConfig>?, previousVersion: Int) {
+    val maxVersion = migrationConfigs?.maxOfOrNull { it.version } ?: previousVersion
```
[the remark this answers is not in the document]
Comment (ellykits) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/DataMigration.kt:126]: By the time we reach this line of code, we are certain there are new migrations. Before this function is called we retrieve new migrations via 

```kotlin 
val previousVersion = preferenceDataStore.read(PreferenceDataStore.MIGRATION_VERSION).first()
val newMigrations = migrations?.filter { it.version > previousVersion }
```

The maximum version is derived from the list of the `newMigrations`

**Comment by ellykits** [comment]
Pushing the UI updates in a few...

**Pull request merged**